src/engine/Server.cpp

```diff
@@ -833,6 +853,61 @@ Awaitable<void> Server::processQuery(
   co_return;
 }
 
+// ____________________________________________________________________________
+Awaitable<void> Server::processUpdate(
+    const ad_utility::url_parser::ParamValueMap& params, const string& update,
+    ad_utility::Timer& requestTimer,
+    const ad_utility::httpUtils::HttpRequest auto& request, auto&& send,
+    TimeLimit timeLimit) {
+  auto messageSender = createMessageSender(queryHub_, request, update);
+
+  auto [cancellationHandle, cancelTimeoutOnDestruction] =
+      setupCancellationHandle(messageSender.getQueryId(), timeLimit);
+
+  auto [pinSubtrees, pinResult] = determineResultPinning(params);
+  LOG(INFO) << "Processing the following SPARQL update:"
+            << (pinResult ? " [pin result]" : "")
+            << (pinSubtrees ? " [pin subresults]" : "") << "\n"
+            << update << std::endl;
+  QueryExecutionContext qec(index_, &cache_, allocator_,
+                            sortPerformanceEstimator_, std::ref(messageSender),
+                            pinSubtrees, pinResult);
+  auto plannedQuery = co_await setupPlannedQuery(
+      params, update, qec, cancellationHandle, timeLimit, requestTimer);
+  auto qet = plannedQuery.queryExecutionTree_;
+
+  if (!plannedQuery.parsedQuery_.hasUpdateClause()) {
+    throw std::runtime_error(
+        absl::StrCat("SPARQL UPDATE was request via the HTTP request, but the "
+                     "following query was sent instead of an update: ",
+                     plannedQuery.parsedQuery_._originalString));
+  }
+
+  // Update the delta triples.
+  //
+  // TODO: This does not yet handle concurrent updates correctly, but it should
+  // work fine when a new update is sent only after the previous one has
+  // finished.
+  auto& deltaTriplesManager = index_.deltaTriplesManager();
+  deltaTriplesManager.modify(
+      [this, &plannedQuery, &qet, &cancellationHandle](auto& deltaTriples) {
+        ExecuteUpdate::executeUpdate(index_, plannedQuery.parsedQuery_, qet,
+                                     deltaTriples, cancellationHandle);
+      });
+
+  LOG(INFO) << "Done processing update"
+            << ", total time was " << requestTimer.msecs().count() << " ms"
+            << std::endl;
+  LOG(DEBUG) << "Runtime Info:\n"
+             << qet.getRootOperation()->runtimeInfo().toString() << std::endl;
+  // TODO<qup42> send a proper response
+  // SPARQL 1.1 Protocol 2.2.4 Successful Responses: "The response body of a
+  // successful update request is implementation defined."
```

Comment:
This can be part of a separate PR.

Reply:
Yes, that is also current plan (see [here](https://github.com/ad-freiburg/qlever/pull/1608#issuecomment-2468107720)).